File: cpp/fx-var-engine/src/book.cpp
#include "fxvar/book.hpp"

#include <algorithm>
#include <cctype>
#include <cmath>
#include <set>
#include <sstream>
#include <stdexcept>

namespace fxvar {

namespace {

std::string upper(std::string s) {
  for (char& c : s) c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
  return s;
}

}  // namespace

PairLegs split_pair(const std::string& pair) {
  if (pair.size() != 6 ||
      !std::all_of(pair.begin(), pair.end(), [](unsigned char c) {
        return std::isalpha(c) != 0;
      })) {
    throw std::invalid_argument("FX pair must be 6 letters like 'EURUSD', got '" +
                                pair + "'");
  }
  PairLegs legs{upper(pair.substr(0, 3)), upper(pair.substr(3, 3))};
  if (legs.base == legs.quote)
    throw std::invalid_argument("FX pair has identical legs: '" + pair + "'");
  return legs;
}

std::string fx_factor(const std::string& ccy) {
  const std::string c = upper(ccy);
  if (c == "USD")
    throw std::invalid_argument(
        "USD has no FX factor: its USD price is identically 1");
  return "FX:" + c;
}

std::string ir_factor(const std::string& ccy) { return "IR:" + upper(ccy); }

// ----------------------------------------------------------------- Market
Market::Market(std::map<std::string, double> spot_usd,
               std::map<std::string, double> rates) {
  for (auto& [c, s] : spot_usd) {
    if (!std::isfinite(s) || s <= 0.0) {
      std::ostringstream msg;
      msg << "spot_usd[" << c << "] must be a positive number, got " << s;
      throw std::invalid_argument(msg.str());
    }
    spot_usd_[upper(c)] = s;
  }
  if (auto it = spot_usd_.find("USD"); it != spot_usd_.end()) {
    if (std::abs(it->second - 1.0) > 1e-12)
      throw std::invalid_argument("spot_usd['USD'] must be 1.0 (USD per USD)");
  }
  spot_usd_["USD"] = 1.0;
  for (auto& [c, r] : rates) rates_[upper(c)] = r;
}

double Market::spot(const std::string& ccy) const {
  const auto it = spot_usd_.find(upper(ccy));
  if (it == spot_usd_.end())
    throw std::invalid_argument("no USD spot for currency '" + ccy +
                                "' in Market");
  return it->second;
}

double Market::rate(const std::string& ccy) const {
  const auto it = rates_.find(upper(ccy));
  if (it == rates_.end())
    throw std::invalid_argument("no interest rate for currency '" + ccy +
                                "' in Market");
  return it->second;
}

double Market::cross(const std::string& pair) const {
  const PairLegs legs = split_pair(pair);
  return spot(legs.base) / spot(legs.quote);
}

double Market::forward(const std::string& pair, double expiry) const {
  const PairLegs legs = split_pair(pair);
  return cross(pair) * std::exp((rate(legs.quote) - rate(legs.base)) * expiry);
}

// ------------------------------------------------------------------- Book
namespace {

void validate_position(const Position& p) {
  if (const auto* cash = std::get_if<CashPosition>(&p)) {
    if (cash->ccy.size() != 3)
      throw std::invalid_argument("cash ccy must be a 3-letter code, got '" +
                                  cash->ccy + "'");
  } else if (const auto* spot = std::get_if<SpotPosition>(&p)) {
    split_pair(spot->pair);
    if (spot->entry_rate && *spot->entry_rate <= 0.0)
      throw std::invalid_argument("entry_rate must be > 0");
  } else if (const auto* fwd = std::get_if<ForwardPosition>(&p)) {
    split_pair(fwd->pair);
    if (fwd->expiry < 0.0) throw std::invalid_argument("expiry must be >= 0");
    if (fwd->strike && *fwd->strike <= 0.0)
      throw std::invalid_argument("strike must be > 0");
  }
}

std::set<std::string> position_currencies(const Position& p) {
  if (const auto* cash = std::get_if<CashPosition>(&p))
    return {upper(std::string(cash->ccy))};
  if (const auto* spot = std::get_if<SpotPosition>(&p)) {
    const PairLegs legs = split_pair(spot->pair);
    return {legs.base, legs.quote};
  }
  const auto& fwd = std::get<ForwardPosition>(p);
  const PairLegs legs = split_pair(fwd.pair);
  return {legs.base, legs.quote};
}

std::string upper_copy(const std::string& s) { return upper(s); }

}  // namespace

Book::Book(std::vector<Position> positions, std::string base)
    : positions_(std::move(positions)), base_(upper_copy(base)) {
  for (const auto& p : positions_) validate_position(p);
}

std::vector<std::string> Book::currencies() const {
  std::set<std::string> ccys{base_};
  for (const auto& p : positions_) {
    const auto pc = position_currencies(p);
    ccys.insert(pc.begin(), pc.end());
  }
  return {ccys.begin(), ccys.end()};
}

std::vector<std::string> Book::factors() const {
  std::set<std::string> fx;
  std::set<std::string> ir;
  for (const auto& c : currencies())
    if (c != "USD") fx.insert(fx_factor(c));
  for (const auto& p : positions_) {
    if (const auto* fwd = std::get_if<ForwardPosition>(&p)) {
      const PairLegs legs = split_pair(fwd->pair);
      ir.insert(ir_factor(legs.base));
      ir.insert(ir_factor(legs.quote));
    }
  }
  std::vector<std::string> out(fx.begin(), fx.end());
  out.insert(out.end(), ir.begin(), ir.end());
  return out;
}

// ----------------------------------------------------------- CompiledBook
CompiledBook::CompiledBook(const Book& book, const Market& market) {
  if (book.empty())
    throw std::invalid_argument(
        "cannot compile an empty book: add positions before running VaR");
  factors_ = book.factors();
  base_ = book.base();
  s0_base_ = market.spot(base_);

  auto factor_index = [this](const std::string& f) -> int {
    for (std::size_t j = 0; j < factors_.size(); ++j)
      if (factors_[j] == f) return static_cast<int>(j);
    return -1;
  };
  base_fx_idx_ = (base_ == "USD") ? -1 : factor_index(fx_factor(base_));

  auto add_leg = [&](double amount, const std::string& ccy, double rate0,
                     double expiry) {
    const double df = (expiry > 0.0) ? std::exp(-rate0 * expiry) : 1.0;
    value0_.push_back(amount * market.spot(ccy) * df);
    fx_idx_.push_back(ccy == "USD" ? -1 : factor_index(fx_factor(ccy)));
    ir_idx_.push_back(expiry > 0.0 ? factor_index(ir_factor(ccy)) : -1);
    neg_expiry_.push_back(expiry > 0.0 ? -expiry : 0.0);
  };

  for (const auto& p : book.positions()) {
    if (const auto* cash = std::get_if<CashPosition>(&p)) {
      add_leg(cash->amount, upper_copy(cash->ccy), 0.0, 0.0);
    } else if (const auto* spot = std::get_if<SpotPosition>(&p)) {
      const PairLegs legs = split_pair(spot->pair);
      const double x0 =
          spot->entry_rate ? *spot->entry_rate : market.cross(spot->pair);
      add_leg(+spot->notional, legs.base, 0.0, 0.0);
      add_leg(-spot->notional * x0, legs.quote, 0.0, 0.0);
    } else {
      const auto& fwd = std::get<ForwardPosition>(p);
      const PairLegs legs = split_pair(fwd.pair);
      const double k =
          fwd.strike ? *fwd.strike : market.forward(fwd.pair, fwd.expiry);
      if (fwd.expiry > 0.0) {
        add_leg(+fwd.notional, legs.base, market.rate(legs.base), fwd.expiry);
        add_leg(-fwd.notional * k, legs.quote, market.rate(legs.quote),
                fwd.expiry);
      } else {  // expired forward = spot difference vs strike
        add_leg(+fwd.notional, legs.base, 0.0, 0.0);
        add_leg(-fwd.notional * k, legs.quote, 0.0, 0.0);
      }
    }
  }
  v0_usd_ = 0.0;
  for (double v : value0_) v0_usd_ += v;
}

double CompiledBook::value_usd(const double* shocks) const {
  double total = 0.0;
  const std::size_t n = value0_.size();
  for (std::size_t i = 0; i < n; ++i) {
    double shift = 0.0;
    if (fx_idx_[i] >= 0) shift += shocks[fx_idx_[i]];
    if (ir_idx_[i] >= 0) shift += neg_expiry_[i] * shocks[ir_idx_[i]];
    total += (shift == 0.0) ? value0_[i] : value0_[i] * std::exp(shift);
  }
  return total;
}

double CompiledBook::pnl(const double* shocks) const {
  const double v1 = value_usd(shocks);
  const double s1_base =
      (base_fx_idx_ >= 0) ? s0_base_ * std::exp(shocks[base_fx_idx_]) : s0_base_;
  return v1 / s1_base - v0_usd_ / s0_base_;
}

double CompiledBook::pnl(const std::map<std::string, double>& shocks) const {
  if (shocks.count("FX:USD"))
    throw std::invalid_argument(
        "shock to 'FX:USD' is not a valid factor: USD is the pivot (its USD "
        "price is identically 1). Shock the other leg(s).");
  std::vector<double> aligned(factors_.size(), 0.0);
  for (std::size_t j = 0; j < factors_.size(); ++j) {
    const auto it = shocks.find(factors_[j]);
    if (it != shocks.end()) aligned[j] = it->second;
  }
  return pnl(aligned.data());
}

std::vector<double> CompiledBook::pnl(const ReturnsMatrix& scenarios) const {
  // Map each leg's factor to a scenario column once, then stream rows.
  std::vector<int> fx_col(value0_.size(), -1), ir_col(value0_.size(), -1);
  std::vector<std::string> missing;
  std::vector<int> factor_col(factors_.size(), -1);
  for (std::size_t j = 0; j < factors_.size(); ++j) {
    factor_col[j] = scenarios.column_index(factors_[j]);
    if (factor_col[j] < 0) missing.push_back(factors_[j]);
  }
  if (!missing.empty()) {
    std::ostringstream msg;
    msg << "scenario matrix is missing required factor columns:";
    for (const auto& f : missing) msg << " " << f;
    throw std::invalid_argument(msg.str());
  }
  for (std::size_t i = 0; i < value0_.size(); ++i) {
    if (fx_idx_[i] >= 0) fx_col[i] = factor_col[fx_idx_[i]];
    if (ir_idx_[i] >= 0) ir_col[i] = factor_col[ir_idx_[i]];
  }
  const int base_col = (base_fx_idx_ >= 0) ? factor_col[base_fx_idx_] : -1;

  const std::size_t n_scen = scenarios.n_obs();
  const std::size_t n_legs = value0_.size();
  std::vector<double> out(n_scen);
  const double v0_base = v0_usd_ / s0_base_;
  for (std::size_t s = 0; s < n_scen; ++s) {
    const double* row = scenarios.data.row(s);
    double v1 = 0.0;
    for (std::size_t i = 0; i < n_legs; ++i) {
      double shift = 0.0;
      if (fx_col[i] >= 0) shift += row[fx_col[i]];
      if (ir_col[i] >= 0) shift += neg_expiry_[i] * row[ir_col[i]];
      v1 += (shift == 0.0) ? value0_[i] : value0_[i] * std::exp(shift);
    }
    const double s1_base =
        (base_col >= 0) ? s0_base_ * std::exp(row[base_col]) : s0_base_;
    out[s] = v1 / s1_base - v0_base;
  }
  return out;
}

std::vector<double> CompiledBook::linear_exposures(double bump) const {
  if (!(bump > 0.0)) throw std::invalid_argument("bump must be > 0");
  std::vector<double> w(factors_.size(), 0.0);
  std::vector<double> shocks(factors_.size(), 0.0);
  for (std::size_t j = 0; j < factors_.size(); ++j) {
    shocks[j] = bump;
    const double up = pnl(shocks.data());
    shocks[j] = -bump;
    const double dn = pnl(shocks.data());
    shocks[j] = 0.0;
    w[j] = (up - dn) / (2.0 * bump);
  }
  return w;
}

}  // namespace fxvar

File: python/fx/03-var-es-engine/src/fx_var/book.py
"""Multi-currency FX book: positions, base-currency P&L, triangulation.

Every currency is represented by its USD factor (USD price of 1 unit, see
:mod:`fx_var.common`).  A position in a *cross* pair (EURJPY) is decomposed
into its two USD legs - long EUR leg, short JPY leg - so cross risk is
triangulated by construction and the book's P&L in any base currency is
consistent with the USD-pair factor set.

Position types
--------------
``Cash``     : a currency balance.  Riskless when denominated in the book's
               base currency.
``Spot``     : a spot FX position, long ``notional`` of the pair's base ccy
               against the quote ccy, struck at ``entry_rate`` (defaults to
               the reference market's cross rate, i.e. zero initial value).
``Forward``  : an outright forward, represented as spot plus two deposit
               legs:  +N e^{-r_f T} of base ccy and -N K e^{-r_d T} of quote
               ccy.  Forward points therefore expose the position to both
               currencies' interest-rate factors (CIP-consistent
               revaluation; tested in tests/test_forwards.py).
``Option``   : a European FX option revalued with Garman-Kohlhagen, either
               by full revaluation or a delta-vega(-gamma) mapping.

P&L convention
--------------
``Book.pnl`` returns profit (+) / loss (-) in the book's **base currency**:
``PnL = V1_usd / S1_base - V0_usd / S0_base`` where the base currency's own
USD price is shocked consistently.  A pure base-ccy cash balance therefore
carries exactly zero risk.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Mapping, Sequence, Union

import numpy as np
import pandas as pd

from .common import fx_factor, ir_factor, split_pair, validate_finite, vol_factor
from .gk import gk_delta, gk_gamma, gk_price, gk_vega

__all__ = ["Market", "Cash", "Spot", "Forward", "Option", "Book", "Position"]

ShockLike = Union[Mapping[str, "float | np.ndarray"], pd.Series, pd.DataFrame]


# --------------------------------------------------------------------------
# Market snapshot
# --------------------------------------------------------------------------
@dataclass(frozen=True)
class Market:
    """Point-in-time market snapshot.

    Parameters
    ----------
    spot_usd : mapping ccy -> float
        USD price of 1 unit of each currency (``spot_usd["EUR"] = 1.08``
        means EURUSD = 1.08).  ``"USD"`` is implied at 1.0 and may be
        omitted; if present it must equal 1.0.
    rates : mapping ccy -> float
        Continuously compounded zero rates, annualised, ACT/365 (flat curve
        per currency - adequate at VaR granularity, see METHODOLOGY.md A7).
    vols : mapping pair -> float
        Annualised lognormal implied vol per option pair, e.g.
        ``{"EURUSD": 0.075}``.
    """

    spot_usd: Mapping[str, float]
    rates: Mapping[str, float] = field(default_factory=dict)
    vols: Mapping[str, float] = field(default_factory=dict)

    def __post_init__(self) -> None:
        spots = dict(self.spot_usd)
        if "USD" in spots and abs(spots["USD"] - 1.0) > 1e-12:
            raise ValueError("spot_usd['USD'] must be 1.0 (USD per USD)")
        spots.setdefault("USD", 1.0)
        for c, s in spots.items():
            if not np.isfinite(s) or s <= 0:
                raise ValueError(f"spot_usd[{c!r}] must be a positive number, got {s}")
        rates = dict(self.rates)
        for c, r in rates.items():
            # NaN policy: refuse.  A NaN rate otherwise propagates silently
            # into a NaN P&L and a NaN VaR.
            if not np.isfinite(r):
                raise ValueError(
                    f"rates[{c!r}] must be a finite number, got {r} "
                    "(NaN policy: refuse, never impute)"
                )
        vols = {k.upper(): v for k, v in self.vols.items()}
        for p, v in vols.items():
            # A NaN vol is *especially* dangerous here: the Garman-Kohlhagen
            # degenerate-limit branch treats a non-finite sigma*sqrt(T) as the
            # zero-vol case and silently returns forward intrinsic instead of
            # a price.  Refuse it at the boundary.
            if not np.isfinite(v):
                raise ValueError(
                    f"vols[{p!r}] must be a finite number, got {v} "
                    "(NaN policy: refuse, never impute)"
                )
            if v < 0:
                raise ValueError(f"vols[{p!r}] must be non-negative, got {v}")
        object.__setattr__(self, "spot_usd", spots)
        object.__setattr__(self, "rates", rates)
        object.__setattr__(self, "vols", vols)

    def spot(self, ccy: str) -> float:
        """USD price of 1 unit of ``ccy``."""
        try:
            return self.spot_usd[ccy]
        except KeyError:
            raise KeyError(f"no USD spot for currency {ccy!r} in Market") from None

    def rate(self, ccy: str) -> float:
        """cc zero rate for ``ccy`` (ACT/365, annualised)."""
        try:
            return self.rates[ccy]
        except KeyError:
            raise KeyError(f"no interest rate for currency {ccy!r} in Market") from None

    def cross(self, pair: str) -> float:
        """Cross rate of ``pair`` (QUOTE per 1 BASE) by USD triangulation."""
        base, quote = split_pair(pair)
        return self.spot(base) / self.spot(quote)

    def forward(self, pair: str, expiry: float) -> float:
        """CIP forward: ``F = X * exp((r_d - r_f) * T)`` (QUOTE per BASE)."""
        base, quote = split_pair(pair)
        return self.cross(pair) * np.exp((self.rate(quote) - self.rate(base)) * expiry)

    def vol(self, pair: str) -> float:
        """Annualised implied vol for ``pair``."""
        try:
            return self.vols[pair.upper()]
        except KeyError:
            raise KeyError(f"no implied vol for pair {pair!r} in Market") from None


# --------------------------------------------------------------------------
# Positions
# --------------------------------------------------------------------------
@dataclass(frozen=True)
class Cash:
    """A cash balance of ``amount`` units of ``ccy``."""

    ccy: str
    amount: float

    def __post_init__(self) -> None:
        validate_finite(amount=self.amount)

    def currencies(self) -> set[str]:
        return {self.ccy.upper()}

    def rate_currencies(self) -> set[str]:
        return set()

    def vol_pairs(self) -> set[str]:
        return set()


@dataclass(frozen=True)
class Spot:
    """Spot FX position: long ``notional`` of BASE ccy vs QUOTE at ``entry_rate``.

    ``notional`` is in units of the pair's base currency; negative = short.
    ``entry_rate=None`` means struck at the reference market's cross rate,
    i.e. zero initial value.  Internally the position is two cash legs
    (+N BASE, -N*X0 QUOTE) - the USD triangulation of the cross.
    """

    pair: str
    notional: float
    entry_rate: float | None = None

    def __post_init__(self) -> None:
        split_pair(self.pair)  # validate
        validate_finite(notional=self.notional, entry_rate=self.entry_rate)
        if self.entry_rate is not None and self.entry_rate <= 0:
            raise ValueError("entry_rate must be > 0")

    def currencies(self) -> set[str]:
        return set(split_pair(self.pair))

    def rate_currencies(self) -> set[str]:
        return set()

    def vol_pairs(self) -> set[str]:
        return set()


@dataclass(frozen=True)
class Forward:
    """Outright FX forward: long ``notional`` BASE at strike ``strike``, expiry ``expiry`` (years).

    Represented as spot + two deposit legs (CIP): value in USD is
    ``N e^{-r_f T} S_base_usd - N K e^{-r_d T} S_quote_usd``, so the position
    responds to both FX factors and both IR factors (forward-point risk).
    ``strike=None`` resolves to the ATM CIP forward of the reference market.
    """

    pair: str
    notional: float
    expiry: float
    strike: float | None = None

    def __post_init__(self) -> None:
        split_pair(self.pair)
        validate_finite(notional=self.notional, expiry=self.expiry,
                        strike=self.strike)
        if self.expiry < 0:
            raise ValueError("expiry must be >= 0")
        if self.strike is not None and self.strike <= 0:
            raise ValueError("strike must be > 0")

    def currencies(self) -> set[str]:
        return set(split_pair(self.pair))

    def rate_currencies(self) -> set[str]:
        return set(split_pair(self.pair))

    def vol_pairs(self) -> set[str]:
        return set()


@dataclass(frozen=True)
class Option:
    """European FX option (Garman-Kohlhagen), ``notional`` in BASE ccy units.

    ``kind='call'`` is a call on the BASE currency (right to buy BASE at K
    units of QUOTE).  Premium/value is carried in QUOTE ccy per unit BASE
    notional and converted to USD/base via the quote leg.  Implied vol is
    read from ``Market.vols[pair]`` and shocked via the ``VOL:PAIR`` factor.
    """

    pair: str
    notional: float
    strike: float
    expiry: float
    kind: str = "call"

    def __post_init__(self) -> None:
        split_pair(self.pair)
        validate_finite(notional=self.notional, strike=self.strike,
                        expiry=self.expiry)
        if self.strike <= 0:
            raise ValueError("strike must be > 0")
        if self.expiry < 0:
            raise ValueError("expiry must be >= 0")
        if self.kind not in ("call", "put"):
            raise ValueError(f"kind must be 'call' or 'put', got {self.kind!r}")

    def currencies(self) -> set[str]:
        return set(split_pair(self.pair))

    def rate_currencies(self) -> set[str]:
        return set(split_pair(self.pair))

    def vol_pairs(self) -> set[str]:
        return {self.pair.upper()}


Position = Union[Cash, Spot, Forward, Option]


# --------------------------------------------------------------------------
# Book
# --------------------------------------------------------------------------
class Book:
    """A multi-currency book with a designated base (reporting) currency.

    Parameters
    ----------
    positions : sequence of Position
        Cash / Spot / Forward / Option positions (may be empty).
    base : str
        Reporting currency for P&L (default ``"USD"``).
    """

    def __init__(self, positions: Sequence[Position] = (), base: str = "USD") -> None:
        self.positions: list[Position] = list(positions)
        self.base = base.upper()

    # -- introspection ----------------------------------------------------
    def currencies(self) -> set[str]:
        """All currencies the book touches, including the base currency."""
        ccys = {self.base}
        for p in self.positions:
            ccys |= p.currencies()
        return ccys

    def factors(self, market: Market | None = None) -> list[str]:
        """Sorted risk-factor names the book is exposed to.

        FX factors for every non-USD currency involved (incl. base if not
        USD), IR factors for forward/option leg currencies, VOL factors for
        option pairs.
        """
        fx = {fx_factor(c) for c in self.currencies() if c != "USD"}
        ir: set[str] = set()
        vol: set[str] = set()
        for p in self.positions:
            ir |= {ir_factor(c) for c in p.rate_currencies()}
            vol |= {vol_factor(q) for q in p.vol_pairs()}
        return sorted(fx) + sorted(ir) + sorted(vol)

    # -- valuation --------------------------------------------------------
    @staticmethod
    def _normalise_shocks(shocks: ShockLike | None) -> dict[str, np.ndarray | float]:
        if shocks is None:
            return {}
        if isinstance(shocks, pd.DataFrame):
            out: dict[str, np.ndarray | float] = {
                c: shocks[c].to_numpy(dtype=float) for c in shocks.columns
            }
        elif isinstance(shocks, pd.Series):
            out = {str(k): float(v) for k, v in shocks.items()}
        else:
            out = {
                str(k): (np.asarray(v, dtype=float) if np.ndim(v) else float(v))
                for k, v in shocks.items()
            }
        if "FX:USD" in out:
            raise ValueError(
                "shock to 'FX:USD' is not a valid factor: USD is the pivot "
                "(its USD price is identically 1). Shock the other leg(s)."
            )
        # NaN policy: refuse.  A single NaN scenario entry otherwise
        # propagates to a NaN P&L, a NaN quantile and a NaN VaR without any
        # exception being raised anywhere on the path.
        for name, value in out.items():
            if not np.all(np.isfinite(np.asarray(value, dtype=float))):
                raise ValueError(
                    f"shock for factor {name!r} contains NaN or infinite "
                    "values (NaN policy: refuse, never impute)"
                )
        return out

    def _shocked_curves(
        self, market: Market, shocks: dict[str, np.ndarray | float]
    ):
        """Build shocked spot/rate/vol lookups (values broadcastable)."""
        ccys = self.currencies()
        pairs: set[str] = set()
        rate_ccys: set[str] = set()
        for p in self.positions:
            pairs |= p.vol_pairs()
            rate_ccys |= p.rate_currencies()
        spot = {}
        for c in ccys:
            if c == "USD":
                spot[c] = 1.0
                continue
            s0 = market.spot(c)
            spot[c] = s0 * np.exp(shocks.get(fx_factor(c), 0.0))
        rates = {c: market.rate(c) + shocks.get(ir_factor(c), 0.0) for c in rate_ccys}
        vols = {q: market.vol(q) + shocks.get(vol_factor(q), 0.0) for q in pairs}
        for q, v in vols.items():
            if np.any(np.asarray(v) < 0):
                raise ValueError(
                    f"shocked implied vol for {q} is negative; cap the vol "
                    "shock or use a floored scenario"
                )
        return spot, rates, vols

    def _position_value_usd(self, pos, spot, rates, vols, ref: Market, option_method: str):
        if isinstance(pos, Cash):
            return pos.amount * spot[pos.ccy.upper()]
        b, q = split_pair(pos.pair)
        if isinstance(pos, Spot):
            x0 = pos.entry_rate if pos.entry_rate is not None else ref.cross(pos.pair)
            return pos.notional * spot[b] - pos.notional * x0 * spot[q]
        if isinstance(pos, Forward):
            k = pos.strike if pos.strike is not None else ref.forward(pos.pair, pos.expiry)
            df_f = np.exp(-rates[b] * pos.expiry)
            df_d = np.exp(-rates[q] * pos.expiry)
            return pos.notional * df_f * spot[b] - pos.notional * k * df_d * spot[q]
        if isinstance(pos, Option):
            x = spot[b] / spot[q]
            sigma = vols[pos.pair.upper()]
            if option_method == "full":
                price_q = gk_price(x, pos.strike, pos.expiry, rates[q], rates[b], sigma, pos.kind)
                return pos.notional * price_q * spot[q]
            # delta-vega(-gamma) mapping around the reference market
            x0 = ref.cross(pos.pair)
            s0 = ref.vol(pos.pair)
            rd0, rf0 = ref.rate(q), ref.rate(b)
            p0 = gk_price(x0, pos.strike, pos.expiry, rd0, rf0, s0, pos.kind)
            delta = gk_delta(x0, pos.strike, pos.expiry, rd0, rf0, s0, pos.kind)
            vega = gk_vega(x0, pos.strike, pos.expiry, rd0, rf0, s0)
            dx = x - x0
            dsig = sigma - s0
            dp = delta * dx + vega * dsig
            if option_method == "delta_vega_gamma":
                gamma = gk_gamma(x0, pos.strike, pos.expiry, rd0, rf0, s0)
                dp = dp + 0.5 * gamma * dx**2
            elif option_method != "delta_vega":
                raise ValueError(
                    "option_method must be 'full', 'delta_vega' or "
                    f"'delta_vega_gamma', got {option_method!r}"
                )
            return pos.notional * (p0 + dp) * spot[q]
        raise TypeError(f"unknown position type {type(pos).__name__}")

    def value_usd(self, market: Market, shocks: ShockLike | None = None,
                  option_method: str = "full"):
        """Book value in USD at ``market`` after applying ``shocks``.

        Broadcasts over scenario arrays in ``shocks``; scalar in/scalar out.
        """
        if option_method not in ("full", "delta_vega", "delta_vega_gamma"):
            # Validated up front so a typo is caught even on option-free
            # books (where the per-position branch would never see it).
            raise ValueError(
                "option_method must be 'full', 'delta_vega' or "
                f"'delta_vega_gamma', got {option_method!r}"
            )
        sh = self._normalise_shocks(shocks)
        spot, rates, vols = self._shocked_curves(market, sh)
        total = 0.0
        for p in self.positions:
            total = total + self._position_value_usd(p, spot, rates, vols, market, option_method)
        # empty book / all-scalar path returns a python float
        return total

    def pnl(self, market: Market, shocks: ShockLike, option_method: str = "full"):
        """Base-currency P&L of the book under factor ``shocks``.

        Parameters
        ----------
        market : Market
            Reference (pre-shock) market snapshot.
        shocks : mapping / Series / DataFrame
            Factor shocks: ``FX:*`` are log returns of CCYUSD, ``IR:*`` and
            ``VOL:*`` absolute changes.  Array-valued entries broadcast to
            scenario vectors; a DataFrame gives one scenario per row.
        option_method : {"full", "delta_vega", "delta_vega_gamma"}
            Option revaluation: full Garman-Kohlhagen repricing or Greek
            mapping around the reference market.

        Returns
        -------
        float or numpy.ndarray
            Profit (+) / loss (-) in the base currency, one value per
            scenario.
        """
        sh = self._normalise_shocks(shocks)
        v0 = self.value_usd(market, None, option_method="full")
        v1 = self.value_usd(market, sh, option_method=option_method)
        if self.base == "USD":
            s0b, s1b = 1.0, 1.0
        else:
            s0b = market.spot(self.base)
            s1b = s0b * np.exp(sh.get(fx_factor(self.base), 0.0))
        out = v1 / s1b - v0 / s0b
        if np.ndim(out) == 0:
            return float(out)
        return np.asarray(out, dtype=float)

    def linear_exposures(self, market: Market, factors: Sequence[str] | None = None,
                         bump: float = 1e-6) -> pd.Series:
        """Delta exposures dPnL/dfactor by central finite differences.

        Units: base-ccy P&L per unit factor move - for ``FX:*`` per unit log
        return (i.e. base-ccy notional exposure), for ``IR:*`` per 1.00 of
        rate (divide by 1e4 for a DV01), for ``VOL:*`` per 1.00 of vol.

        These are the mapping weights used by the variance-covariance method
        and the reverse-stress closed form.
        """
        if factors is None:
            factors = self.factors(market)
        vals = {}
        for f in factors:
            up = self.pnl(market, {f: bump})
            dn = self.pnl(market, {f: -bump})
            vals[f] = (up - dn) / (2.0 * bump)
        return pd.Series(vals, dtype=float).reindex(list(factors))
